File: pyspider/browser/browser.py
"""
浏览器自动化模块
支持 Selenium 和 Playwright
"""

from typing import Optional, Dict, List
from selenium import webdriver
from selenium.webdriver.chrome.options import Options as ChromeOptions
from selenium.webdriver.chrome.service import Service
from webdriver_manager.chrome import ChromeDriverManager
import time


class BrowserManager:
    """浏览器管理器（Selenium）"""
    
    def __init__(self, headless: bool = True, user_agent: Optional[str] = None):
        self.headless = headless
        self.user_agent = user_agent
        self.driver = self._create_driver()
    
    def _create_driver(self):
        """创建浏览器驱动"""
        options = ChromeOptions()
        
        if self.headless:
            options.add_argument("--headless")
        
        options.add_argument("--no-sandbox")
        options.add_argument("--disable-dev-shm-usage")
        options.add_argument("--disable-gpu")
        options.add_argument("--window-size=1920,1080")
        
        if self.user_agent:
            options.add_argument(f"--user-agent={self.user_agent}")
        
        # 隐藏自动化特征
        options.add_experimental_option("excludeSwitches", ["enable-automation"])
        options.add_experimental_option("useAutomationExtension", False)
        
        driver = webdriver.Chrome(
            service=Service(ChromeDriverManager().install()),
            options=options
        )
        
        # 注入脚本隐藏自动化特征
        driver.execute_cdp_cmd("Page.addScriptToEvaluateOnNewDocument", {
            "source": """
                Object.defineProperty(navigator, 'webdriver', {
                    get: () => undefined
                });
            """
        })
        
        return driver
    
    def navigate(self, url: str):
        """导航到 URL"""
        self.driver.get(url)
    
    def get_html(self) -> str:
        """获取 HTML"""
        return self.driver.page_source
    
    def get_text(self) -> str:
        """获取文本"""
        return self.driver.find_element("tag name", "body").text
    
    def get_title(self) -> str:
        """获取标题"""
        return self.driver.title
    
    def click(self, selector: str):
        """点击"""
        element = self.driver.find_element("css selector", selector)
        element.click()
    
    def input_text(self, selector: str, text: str):
        """输入文本"""
        element = self.driver.find_element("css selector", selector)
        element.clear()
        element.send_keys(text)
    
    def screenshot(self, path: str):
        """截图"""
        self.driver.save_screenshot(path)
    
    def execute_script(self, script: str, *args):
        """执行 JavaScript"""
        return self.driver.execute_script(script, *args)
    
    def scroll_to_bottom(self):
        """滚动到底部"""
        self.driver.execute_script("window.scrollTo(0, document.body.scrollHeight);")
    
    def wait_for_element(self, selector: str, timeout: int = 30):
        """等待元素"""
        from selenium.webdriver.support.ui import WebDriverWait
        from selenium.webdriver.support import expected_conditions as EC
        from selenium.webdriver.common.by import By
        
        WebDriverWait(self.driver, timeout).until(
            EC.presence_of_element_located((By.CSS_SELECTOR, selector))
        )
    
    def close(self):
        """关闭浏览器"""
        if self.driver:
            self.driver.quit()


class PlaywrightManager:
    """Playwright 浏览器管理器"""
    
    def __init__(self, headless: bool = True):
        self.headless = headless
        self.browser = None
        self.page = None
        self._init_browser()
    
    def _init_browser(self):
        """初始化浏览器"""
        from playwright.sync_api import sync_playwright
        
        playwright = sync_playwright().start()
        self.browser = playwright.chromium.launch(headless=self.headless)
        self.page = self.browser.new_page()
    
    def navigate(self, url: str):
        """导航"""
        self.page.goto(url)
    
    def get_html(self) -> str:
        """获取 HTML"""
        return self.page.content()
    
    def get_text(self, selector: str = "body") -> str:
        """获取文本"""
        return self.page.text_content(selector)
    
    def click(self, selector: str):
        """点击"""
        self.page.click(selector)
    
    def fill(self, selector: str, value: str):
        """填写"""
        self.page.fill(selector, value)
    
    def screenshot(self, path: str):
        """截图"""
        self.page.screenshot(path=path)
    
    def execute_script(self, script: str):
        """执行 JavaScript"""
        return self.page.evaluate(script)
    
    def wait_for_selector(self, selector: str, timeout: int = 30000):
        """等待元素"""
        self.page.wait_for_selector(selector, timeout=timeout)
    
    def scroll_to_bottom(self):
        """滚动到底部"""
        self.page.evaluate("window.scrollTo(0, document.body.scrollHeight)")
    
    def close(self):
        """关闭"""
        if self.browser:
            self.browser.close()


class JavaScriptExecutor:
    """JavaScript 执行器"""
    
    def __init__(self, driver):
        self.driver = driver
    
    def execute(self, script: str, *args):
        """执行脚本"""
        return self.driver.execute_script(script, *args)
    
    def get_title(self) -> str:
        """获取标题"""
        return self.driver.execute_script("return document.title")
    
    def get_url(self) -> str:
        """获取 URL"""
        return self.driver.execute_script("return window.location.href")
    
    def get_cookies(self) -> Dict[str, str]:
        """获取 Cookie"""
        cookies = {}
        for cookie in self.driver.get_cookies():
            cookies[cookie['name']] = cookie['value']
        return cookies
    
    def set_cookie(self, name: str, value: str):
        """设置 Cookie"""
        self.driver.add_cookie({'name': name, 'value': value})
    
    def get_local_storage(self) -> Dict[str, str]:
        """获取 localStorage"""
        return self.driver.execute_script("""
            var ls = window.localStorage;
            var obj = {};
            for (var i = 0; i < ls.length; i++) {
                obj[ls.key(i)] = ls.getItem(ls.key(i));
            }
            return obj;
        """)
    
    def bypass_detection(self):
        """绕过自动化检测"""
        self.driver.execute_script("""
            Object.defineProperty(navigator, 'webdriver', {
                get: () => undefined
            });
            Object.defineProperty(navigator, 'plugins', {
                get: () => [1, 2, 3, 4, 5]
            });
            Object.defineProperty(navigator, 'languages', {
                get: () => ['zh-CN', 'zh', 'en']
            });
        """)

File: rustspider/src/browser/browser.rs
//! 浏览器自动化模块
//! 使用 fantoccini (WebDriver 客户端)

use std::time::Duration;

/// 浏览器管理器
#[cfg(feature = "browser")]
pub struct BrowserManager {
    webdriver: fantoccini::Client,
}

#[cfg(feature = "browser")]
impl BrowserManager {
    /// 创建浏览器管理器
    pub async fn new(headless: bool) -> Result<Self, fantoccini::error::CmdError> {
        use fantoccini::{ClientBuilder, Locator};
        
        let mut caps = serde_json::map::Map::new();
        
        if headless {
            let mut args = serde_json::map::Map::new();
            args.insert("args".to_string(), vec!["--headless".into()].into());
            caps.insert("goog:chromeOptions".to_string(), args.into());
        }
        
        let webdriver = ClientBuilder::native()
            .capabilities(caps)
            .connect("http://localhost:4444")
            .await?;
        
        Ok(BrowserManager { webdriver })
    }
    
    /// 导航到 URL
    pub async fn navigate(&self, url: &str) -> Result<(), fantoccini::error::CmdError> {
        self.webdriver.goto(url).await
    }
    
    /// 获取 HTML
    pub async fn get_html(&self) -> Result<String, fantoccini::error::CmdError> {
        self.webdriver.source().await
    }
    
    /// 获取标题
    pub async fn get_title(&self) -> Result<String, fantoccini::error::CmdError> {
        self.webdriver.title().await
    }
    
    /// 点击元素
    pub async fn click(&self, selector: &str) -> Result<(), fantoccini::error::CmdError> {
        let element = self.webdriver.wait().for_element(Locator::Css(selector)).await?;
        element.click().await
    }
    
    /// 输入文本
    pub async fn fill(&self, selector: &str, text: &str) -> Result<(), fantoccini::error::CmdError> {
        let element = self.webdriver.wait().for_element(Locator::Css(selector)).await?;
        element.send_keys(text).await
    }
    
    /// 截图
    pub async fn screenshot(&self) -> Result<Vec<u8>, fantoccini::error::CmdError> {
        self.webdriver.screenshot().await
    }
    
    /// 执行 JavaScript
    pub async fn execute_script(&self, script: &str) -> Result<(), fantoccini::error::CmdError> {
        self.webdriver.execute(script, Vec::new()).await
    }
    
    /// 滚动到底部
    pub async fn scroll_to_bottom(&self) -> Result<(), fantoccini::error::CmdError> {
        self.webdriver.execute("window.scrollTo(0, document.body.scrollHeight);", Vec::new()).await
    }
    
    /// 等待元素
    pub async fn wait_for_element(&self, selector: &str, timeout: Duration) -> Result<(), fantoccini::error::CmdError> {
        self.webdriver.wait().with_timeout(timeout).for_element(Locator::Css(selector)).await?;
        Ok(())
    }
    
    /// 关闭浏览器
    pub async fn close(self) -> Result<(), fantoccini::error::CmdError> {
        self.webdriver.close().await
    }
}

/// 滚动加载器
pub struct ScrollLoader {
    scroll_action: Box<dyn FnMut() + Send>,
    get_height_action: Box<dyn FnMut() -> i64 + Send>,
}

impl ScrollLoader {
    /// 创建滚动加载器
    pub fn new<F, G>(scroll_fn: F, get_height_fn: G) -> Self
    where
        F: FnMut() + Send + 'static,
        G: FnMut() -> i64 + Send + 'static,
    {
        ScrollLoader {
            scroll_action: Box::new(scroll_fn),
            get_height_action: Box::new(get_height_fn),
        }
    }
    
    /// 滚动到底部
    pub fn scroll_to_bottom(&mut self, pause_ms: u64, max_scrolls: usize) -> usize {
        use crate::dynamic::wait::DynamicWait;
        
        let mut scroll_count = 0;
        let mut last_height = (self.get_height_action)();
        let mut stable_count = 0;
        
        while scroll_count < max_scrolls {
            // 滚动到底部
            (self.scroll_action)();
            scroll_count += 1;
            
            // 等待加载
            DynamicWait::sleep(pause_ms);
            
            // 检查新高度
            let new_height = (self.get_height_action)();
            
            if new_height == last_height {
                stable_count += 1;
                if stable_count >= 2 {
                    break;
                }
            } else {
                stable_count = 0;
                last_height = new_height;
            }
        }
        
        scroll_count
    }
}

/// JavaScript 执行器
pub struct JavaScriptExecutor<F> {
    execute_fn: F,
}

impl<F> JavaScriptExecutor<F>
where
    F: FnMut(&str) -> Result<serde_json::Value, String>,
{
    /// 创建执行器
    pub fn new(execute_fn: F) -> Self {
        JavaScriptExecutor { execute_fn }
    }
    
    /// 执行脚本
    pub fn execute(&mut self, script: &str) -> Result<serde_json::Value, String> {
        (self.execute_fn)(script)
    }
    
    /// 获取标题
    pub fn get_title(&mut self) -> Result<String, String> {
        let result = self.execute("return document.title")?;
        result.as_str().map(|s| s.to_string()).ok_or_else(|| "Invalid result".to_string())
    }
    
    /// 获取 URL
    pub fn get_url(&mut self) -> Result<String, String> {
        let result = self.execute("return window.location.href")?;
        result.as_str().map(|s| s.to_string()).ok_or_else(|| "Invalid result".to_string())
    }
    
    /// 绕过自动化检测
    pub fn bypass_detection(&mut self) -> Result<(), String> {
        self.execute(r#"
            Object.defineProperty(navigator, 'webdriver', {
                get: () => undefined
            });
            Object.defineProperty(navigator, 'plugins', {
                get: () => [1, 2, 3, 4, 5]
            });
            Object.defineProperty(navigator, 'languages', {
                get: () => ['zh-CN', 'zh', 'en']
            });
        "#)?;
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    
    #[test]
    fn test_scroll_loader() {
        let mut scroll_count = 0;
        let mut height = 1000i64;
        let mut stable = 0;
        
        let mut loader = ScrollLoader::new(
            || { scroll_count += 1; },
            || { 
                if stable < 2 { stable += 1; height } else { height }
            },
        );
        
        let result = loader.scroll_to_bottom(10, 50);
        assert!(result > 0);
    }
}
